src/verifier/step4/SparsePolynomial.sol

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "src/pasta/Vesta.sol";
+import "src/pasta/Pallas.sol";
+
+// Port of Nova' SparsePolynomial (https://github.com/microsoft/Nova/blob/main/src/spartan/polynomial.rs#L121)
+library SparsePolynomialLib {
+    struct Z {
+        uint256 index;
+        uint256 scalar;
+    }
+
+    function evaluate(
+        uint256 num_vars,
+        Z[] memory poly_X,
+        uint256[] memory r_y,
+        uint256 modulus,
+        function (uint256) pure returns (uint256) negateBase
```

Comment:
This is cool! I didn't know this was possible

Reply:
Yeah. I think we can remove some code duplication in functions that just require Pallas / Vesta distinguishing while computation with this approach